### ISSUE TITLE
CheckReturnValue Rule

### ISSUE THREAD

**Issue by vanniktech** [arturbosch/detekt#209]
Given a method that is annotated with `CheckReturnValue` or `CheckResult` and you call it without storing the value it should be flagged. We don't even need to resolve the imports, simply checking for the annotation name is sufficient.

Positive case:

```kotlin
interface Something {
  @CheckReturnValue fun foo(): Int
}

fun something(something: Something) {
  something.foo()
}
```

Negative case:

```kotlin
interface Something {
  @CheckReturnValue fun foo(): Int
}

fun something(something: Something): Int {
  val foo = something.foo()
  return foo + foo
}
```

**Comment by arturbosch** [comment]
I do not think such a rule should be in the default rule set. Who provides this annotation? Should the user create one? Sounds like a custom rule set to me :)

**Comment by vanniktech** [comment]
`@CheckResult` comes from Android and they also have that rule built into lint (that should work with AS 3.0 and forward). `@CheckReturnValue` can be defined, for instance [RxJava 2 has it](https://github.com/ReactiveX/RxJava/blob/46ec6a6365ded7f9d96674baf40f7342d76ebdda/src/main/java/io/reactivex/annotations/CheckReturnValue.java). Also it's part of the JSR 305 annotations.

I'd argue that this rule is pretty common and should be in the main rule set. [ErrorProne](https://github.com/google/error-prone/blob/master/docs/bugpattern/CheckReturnValue.md) has it as well as [Findbugs](http://findbugs.sourceforge.net/api/edu/umd/cs/findbugs/annotations/CheckReturnValue.html) (they also provide their own annotation).

**Comment by arturbosch** [comment]
Ok, sounds important, never used that :). So it tells you that the return value is unused? Like intellij does?

**Comment by vanniktech** [comment]
Exactly yeah. I heavily rely on it when using RxJava (to not forget to call subscribe() and handle the returned Disposable) or custom builders (not calling build). It's just a sanity check.

**Comment by vanniktech** [comment]
Can you point me to the right visit method that I should be using for this?

**Comment by arturbosch** [comment]
We cannot implement this rule right now. My idea would be to visit a `KtNamedFunction`, get the annotation and search for usages of this method. This is the part where resolving must be implemented ^^

**Comment by vanniktech** [comment]
Maybe add a label to the issue and to the other ones where resolving must be implemented first?

**Comment by vanniktech** [comment]
I'm closing this since we'll get this with lint in AGP 3.1.0. Also with 3.1.0 there's a lint plugin that can be used in kotlin code to get the checks. Does not make sense to duplicate the checks and also this is still blocked by type and symbol resolution.
